src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1911 @@
+        r = json_variant_strv(v, &components);
+        if (r < 0)
+                return r;
+
+        if (ret_components)
+                *ret_components = TAKE_PTR(components);
+        if (ret_have_default)
+                *ret_have_default = have_default;
+        return 0;
+}
+
+static int manager_ensure_targets(Manager *m);
+
+static int target_object_find(
+                sd_bus *bus,
+                const char *path,
+                const char *iface,
+                void *userdata,
+                void **found,
+                sd_bus_error *error) {
+
+        Manager *m = ASSERT_PTR(userdata);
+        Target *t;
+        _cleanup_free_ char *e = NULL;
+        const char *p;
+        int r;
+
+        assert(bus);
+        assert(path);
+        assert(found);
+
+        p = startswith(path, "/org/freedesktop/sysupdate1/target/");
+        if (!p)
+                return 0;
+
+        e = bus_label_unescape(p);
+        if (!e)
+                return -ENOMEM;
+
+        r = manager_ensure_targets(m);
+        if (r < 0)
+                return r;
+
+        t = hashmap_get(m->targets, e);
+        if (!t)
+                return 0;
+
+        *found = t;
+        return 1;
+}
+
+static char *target_bus_path(Target *t) {
+        _cleanup_free_ char *e = NULL;
+
+        assert(t);
+
+        e = bus_label_escape(t->id);
+        if (!e)
+                return NULL;
+
+        return strjoin("/org/freedesktop/sysupdate1/target/", e);
+}
+
+static int target_node_enumerator(
+                sd_bus *bus,
+                const char *path,
+                void *userdata,
+                char ***nodes,
+                sd_bus_error *error) {
+
+        _cleanup_strv_free_ char **l = NULL;
+        Manager *m = ASSERT_PTR(userdata);
+        Target *t;
+        unsigned k = 0;
+        int r;
+
+        r = manager_ensure_targets(m);
+        if (r < 0)
+                return r;
+
+        l = new0(char*, hashmap_size(m->targets) + 1);
+        if (!l)
+                return -ENOMEM;
+
+        HASHMAP_FOREACH(t, m->targets) {
+                l[k] = target_bus_path(t);
+                if (!l[k])
+                        return -ENOMEM;
+                k++;
+        }
+
+        *nodes = TAKE_PTR(l);
+        return 1;
+}
+
+static const sd_bus_vtable target_vtable[] = {
+        SD_BUS_VTABLE_START(0),
+
+        SD_BUS_PROPERTY("Class", "s", target_property_get_class,
+                        offsetof(Target, class), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Name", "s", NULL, offsetof(Target, name),
+                        SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Path", "s", NULL, offsetof(Target, path),
+                        SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Version", "s", target_property_get_cached, 0, 0),
+        SD_BUS_PROPERTY("Appstream", "s", target_property_get_cached, 0,
+                        SD_BUS_VTABLE_PROPERTY_CONST),
```

Comment:
I a pretty sure these should not be properties, as they are "heavy" to acquire, and acquiring can fail for too many reasons. D-Bus propertys should never really fail to be acquired. 

i.e. just make this a method GetAppstream()